src/wp-includes/class-wp-token-map.php

```diff
@@ -0,0 +1,810 @@
+	 * @param ?int    $offset            How many bytes into the string where the lookup key ought to start.
+	 * @param ?int    &$skip_bytes       Holds byte-length of found lookup key if matched, otherwise not set.
+	 * @param ?string $case_sensitivity 'case-insensitive' to ignore ASCII case or default of 'case-sensitive'.
+	 * @return string|false Mapped value of lookup key if found, otherwise `false`.
+	 */
+	public function read_token( $text, $offset = 0, &$skip_bytes = null, $case_sensitivity = 'case-sensitive' ) {
+		$ignore_case = 'case-insensitive' === $case_sensitivity;
+		$text_length = strlen( $text );
+
+		// Search for a long word first, if the text is long enough, and if that fails, a short one.
+		if ( $text_length > $this->key_length ) {
+			$group_key = substr( $text, $offset, $this->key_length );
+
+			$group_at = $ignore_case ? stripos( $this->groups, $group_key ) : strpos( $this->groups, $group_key );
+			if ( false === $group_at ) {
+				// Perhaps a short word then.
+				return strlen( $this->small_words ) > 0
+					? $this->read_small_token( $text, $offset, $skip_bytes, $case_sensitivity )
+					: false;
+			}
+
+			$group        = $this->large_words[ $group_at / ( $this->key_length + 1 ) ];
+			$group_length = strlen( $group );
+			$at           = 0;
+			while ( $at < $group_length ) {
+				$token_length   = unpack( 'C', $group[ $at++ ] )[1];
+				$token          = substr( $group, $at, $token_length );
+				$at            += $token_length;
+				$mapping_length = unpack( 'C', $group[ $at++ ] )[1];
+				$mapping_at     = $at;
+
+				if ( 0 === substr_compare( $text, $token, $offset + $this->key_length, $token_length, $ignore_case ) ) {
+					$skip_bytes = $this->key_length + $token_length;
+					return substr( $group, $mapping_at, $mapping_length );
+				}
+
+				$at = $mapping_at + $mapping_length;
+			}
+		}
+
+		// Perhaps a short word then.
+		return strlen( $this->small_words ) > 0
+			? $this->read_small_token( $text, $offset, $skip_bytes, $case_sensitivity )
+			: false;
+	}
+
+	/**
+	 * Finds a match for a short word at the index.
+	 *
+	 * @since 6.6.0.
+	 *
+	 * @param string  $text             String in which to search for a lookup key.
+	 * @param ?int    $offset           How many bytes into the string where the lookup key ought to start.
+	 * @param ?int    &$skip_bytes      Holds byte-length of found lookup key if matched, otherwise not set.
+	 * @param ?string $case_sensitivity 'case-insensitive' to ignore ASCII case or default of 'case-sensitive'.
+	 * @return string|false Mapped value of lookup key if found, otherwise `false`.
+	 */
+	private function read_small_token( $text, $offset, &$skip_bytes, $case_sensitivity = 'case-sensitive' ) {
+		$ignore_case  = 'case-insensitive' === $case_sensitivity;
+		$small_length = strlen( $this->small_words );
+		$search_text  = substr( $text, $offset, $this->key_length );
+		if ( $ignore_case ) {
+			$search_text = strtoupper( $search_text );
+		}
```

Comment:
thanks for the alternative exploration. I am not sure it would be possible to use `strcasecmp` like this here because we aren't looking for full matches, necessarily, in the `small_words` array. we're really looking for a match in the `$text` of a string in the small words array, but the words in the small words array are zero-extended.

for example, if we have `ab` in `$text` at the `$offset`, and `"a\x00\x00"` in the small words array, we will not match `0 === strcasecmp()`, but there is indeed a match.

so we could add a length to the comparison, but then if it didn't match, we'd need to backtrack and try matching with `length - 1` and so on until `length === 1` and there are no matches.

by crawling forward only in the document we can ensure that any prefix of the input text will match as expected on a word shorter than the longest small word.

Reply:
The idea is still to do character by character comparison, not full string comparison, but instead of calling `strtoupper` on the needle and on the characters, `strcasecmp` does the work of ASCII case-insensitive matching.